gateway/src/lib.rs

```diff
@@ -366,24 +367,30 @@ impl Actor {
                         e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "error flushing checkpoint")
                     })?;
 
+                    let cross_msgs = commit.take_cross_msgs();
+
                     // update prev_check for child
                     sub.prev_checkpoint = Some(commit);
                     // flush subnet
                     st.flush_subnet(rt.store(), &sub).map_err(|e| {
                         e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "error flushing subnet")
                     })?;
-                    Ok(fee)
-                }
-                None => {
-                    Err(actor_error!(
-                        illegal_argument,
-                        "subnet with id {} not registered",
-                        shid
-                    ))
+                    Ok((fee, cross_msgs))
                 }
+                None => Err(actor_error!(
+                    illegal_argument,
+                    "subnet with id {} not registered",
+                    shid
+                )),
             }
         })?;
 
+        if let Some(msgs) = cross_msgs {
```

Comment:
Really important, we need to ensure that the messages are ordered by nonce before they are executed